Deepspeed Ulysses/ALST integration (#3817)
huggingface/accelerate: docs/source/concept_guides/context_parallelism.md

To configure the `deepspeed` backend:

```python
parallelism_config = ParallelismConfig(
    cp_backend="deepspeed",
    cp_size=4,
    cp_handler=DeepSpeedContextParallelConfig(
        cp_seq_length=256,
        cp_seq_length_is_variable=True,
        cp_attn_implementation="sdpa",
    ),
)
accelerator = Accelerator(
    ...,
    parallelism_config=parallelism_config,
)
```

- `cp_backend`: set to `deepspeed` here
- `cp_size` is the degree of the sequence parallelism - in the above example it's 4, therefore 4 gpus will be used to process a single batch.
- `cp_seq_length` and `cp_seq_length_is_variable` are used to deal with sequence lengths. If `cp_seq_length_is_variable=True` the backend will work with a sequence length that may change between batches, in which case `cp_seq_length` value can be set to anything divisible by the context parallel degree or not set at all. In this case on every `forward` the sequence variables will be derived from input. If `False` then `seq_length` needs to match the batch's sequence length dimension, which then will have to be padded to be always the same. The default is `True`.

> - `cp_attn_implementation` is one of `sdpa`, `flash_attention_2` or `flash_attention_3`. This sequence parallel implementation uses `position_ids` instead of `attention_mask` therefore `eager` can't work here until it'd support working with `position_ids`. Also please note that `sdpa` doesn't handle correctly multiple samples combined into one, it'd attend to the whole sample as one. If the samples aren't combined `sdpa` will work correctly. Therefore Flash Attention should be the ideal choice as it always works.

Reviewer: ```suggestion
- `cp_attn_implementation` is one of `sdpa`, `flash_attention_2` or `flash_attention_3`. This sequence parallel implementation uses `position_ids` instead of `attention_mask` therefore, `eager` can't work here until it supports working with `position_ids`. Also, please note that `sdpa` doesn't handle multiple samples combined into one correctly; it will attend to the whole sample as one. If the samples aren't combined, `sdpa` will work correctly. Therefore, Flash Attention should be the ideal choice as it always works.
```

num_processes: 2
parallelism_config:
  parallelism_config_dp_replicate_size: 1
  parallelism_config_dp_shard_size: 1
  parallelism_config_tp_size: 1
  parallelism_config_cp_size: 2
  parallelism_config_cp_backend: deepspeed
  parallelism_config_cp_seq_length_is_variable: true
  parallelism_config_cp_attn_implementation: sdpa

```

Ulysses sequence parallelism can be combined with data parallelism:

```python
parallelism_config = ParallelismConfig(
    cp_backend="deepspeed",
    cp_size=2,
    dp_shard_size=2,
    cp_handler=DeepSpeedContextParallelConfig(...),
)
```
Here we use 4 gpus, with 2 sequence parallelism replicas. Deepspeed-ZeRO is what drives the data parallelism.

Please note that a lot of magic is hidden inside [UlyssesSPDataLoaderAdapter](https://github.com/deepspeedai/DeepSpeed/blob/64c0052fa08438b4ecf4cae30af15091a92d2108/deepspeed/runtime/sequence_parallel/ulysses_sp.py#L442). It's used behind the scenes, wrapping your original DataLoader object, but you should be aware of it should you run into any problems. It also automatically injects the correct `shift_labels` into the batch dictionary, before the batch gets sharded across the participating ranks.

> Now the only remaining piece to start using ALST/UlyssesSP is to aggregate the loss across ranks using a differentiable `all_gather` to get the grads right. The following code does it, while also exlcuding any masked out with `-100` tokens, to get the correct average:

Reviewer: ```suggestion
Now the only remaining piece to start using ALST/UlyssesSP is to aggregate the loss across ranks using a differentiable `all_gather` to get the grads right. The following code does it, while also excluding any masked out with `-100` tokens, to get the correct average:
```

- `cp_attn_implementation` is one of `sdpa`, `flash_attention_2` or `flash_attention_3`. This sequence parallel implementation uses `position_ids` instead of `attention_mask` therefore `eager` can't work here until it'd support working with `position_ids`. Also please note that `sdpa` doesn't handle correctly multiple samples combined into one, it'd attend to the whole sample as one. If the samples aren't combined `sdpa` will work correctly. Therefore Flash Attention should be the ideal choice as it always works.

Instead of setting these values in `DeepSpeedContextParallelConfig` object, you can also use the environment variables to accomplish the same - here they are correspondingly to the end of the list above.
- `PARALLELISM_CONFIG_CP_BACKEND`
- `PARALLELISM_CONFIG_CP_SEQ_LENGTH`
- `PARALLELISM_CONFIG_CP_SEQ_LENGTH_IS_VARIABLE`
- `PARALLELISM_CONFIG_CP_ATTN_IMPLEMENTATION`

If not passed in the code `cp_size` can be set via `--parallelism_config_cp_size` CLI argument. SAme for other arguments. You can also do the accelerate config file style config, e.f. for 2 gpus:

> 

Reviewer 2: ```suggestion
If not passed in the code `cp_size` can be set via `--parallelism_config_cp_size` CLI argument. Same for other arguments. You can also do the accelerate config file style config, e.f. for 2 gpus:

```
